test/generaloption.py

```diff
@@ -764,6 +773,106 @@ def raise_error(msg, *args):
+        stderr = self.get_stderr()
+        self.assertTrue(msg in stderr)
+
+    def test_nosuchoption(self):
+        """Test catching of non-existing options."""
+        self._match_testoption1_sysexit(['--nosuchoptiondefinedfoobar'],
+                                        "no such option: --nosuchoptiondefinedfoobar")
+
+    def test_is_value_a_commandline_option(self):
+        """Test ExtOptionParser is_value_a_commandline_option method"""
+        topt = TestOption1(go_args=[])
+        tp = topt.parser
+        fn = tp.is_value_a_commandline_option
+
+        self.assertEqual(tp.ALLOW_OPTION_AS_VALUE, False,
+                         msg="default do not allow value as option")
+        self.assertEqual(tp.ALLOW_DASH_AS_VALUE, False,
+                         msg="default do not allow value starting with -")
+        self.assertEqual(tp.ALLOW_TYPO_AS_VALUE, True,
+                         msg="default do allow value similar to option")
+
+        # fake commandline args
+        # this is purely to test the function, actual usage is in test_option_as_value
+        tp.commandline_arguments = ['--base', '-something', 'base']
+        def t(fail, tpl):
+            for idx, value in enumerate(tp.commandline_arguments):
+                msg = tpl % value
+                res = fn(value, index=idx)
+                if idx in fail:
+                    self.assertFalse(res is None, "failure should not return None for value %s" % value)
+                    self.assertTrue(msg in res, msg='%s in %s' % (msg, res))
+                else:
+                    self.assertTrue(res is None)
+
+        # anything goes
+        tp.ALLOW_OPTION_AS_VALUE = True
+        tp.ALLOW_DASH_AS_VALUE = True
+        tp.ALLOW_TYPO_AS_VALUE = True
+
+        t([-1], 'x%sx') # all ok
+
+        tp.ALLOW_OPTION_AS_VALUE = False
+        t([0], "Value '%s' is also a valid option")
+        tp.ALLOW_OPTION_AS_VALUE = True
+
+        tp.ALLOW_DASH_AS_VALUE = False
+        # options also start with a -
+        t([0, 1], "Value '%s' starts with a '-'")
+        tp.ALLOW_DASH_AS_VALUE = True
+
+        tp.ALLOW_TYPO_AS_VALUE = False
+        # an option is a close match for an option, who'd guessed that...
+        t([0,2], "Value '%s' too close match to option(s) ")
+        tp.ALLOW_TYPO_AS_VALUE = True
+
+
+    def test_option_as_value(self):
+        """Test how valid options being used as values are handled."""
+        # -s requires an argument
+        self._match_testoption1_sysexit(['-b', '-s'],
+                                        "-s option requires an argument")
+
+        # valid ways of specifying '-b' as a value to --store/-s
+        topt = TestOption1(go_args=['--store=-b'])
+        self.assertEqual(topt.options.store, '-b')
+        self.assertFalse(topt.options.base)  # remain False (default value)
+
+        topt = TestOption1(go_args=['-s-b'])  # equivalent to --store=-b
```

Comment:
Hmm, but then the issue remains if you really want to pass a value that starts with `-`.

I guess the only option that we should allow (by default) is `--store=-b`, since then it's clear it's (most likely) not a typo (due to the `--` and `=-` combo).